### ISSUE TITLE
[FEA] Multi-GPU support (single node)

### ISSUE THREAD

**Issue by seunghwak** [rapidsai/rmm#81]
Related issue: https://github.com/rapidsai/rmm/issues/66

**Is your feature request related to a problem? Please describe.**

I wish I could use RMM for a multi-GPU node. However, it may not be possible in the current implementation if I enable pool allocation.

```cpp
 54 // Initialize memory manager state and storage.
 55 rmmError_t rmmInitialize(rmmOptions_t *options)
 56 {
 57     rmm::Manager::getInstance().initialize(options);
 58 
 59     if (rmm::Manager::usePoolAllocator())
 60     {
 61         cnmemDevice_t dev;
 62         RMM_CHECK_CUDA( cudaGetDevice(&(dev.device)) );
 63         // Note: cnmem defaults to half GPU memory
 64         dev.size = rmm::Manager::getOptions().initial_pool_size;
 65         dev.numStreams = 1;
 66         cudaStream_t streams[1]; streams[0] = 0;
 67         dev.streams = streams;
 68         dev.streamSizes = 0;
 69         unsigned flags = rmm::Manager::useManagedMemory() ? CNMEM_FLAGS_MANAGED : 0;
 70         RMM_CHECK_CNMEM( cnmemInit(1, &dev, flags) );
 71     }
 72     return RMM_SUCCESS;
 73 }
```

rmmInitialize() calls cnmemInit in line 70 with numDevices set to 1.

```cpp
1071 cnmemStatus_t cnmemInit(int numDevices, const cnmemDevice_t *devices, unsigned flags) {
1072     // Make sure we have at least one device declared.
1073     CNMEM_CHECK_TRUE(numDevices > 0, CNMEM_STATUS_INVALID_ARGUMENT);
1074 
1075     // Find the largest ID of the device.
1076     int maxDevice = 0;
1077     for( int i = 0 ; i < numDevices ; ++i ) {
1078         if( devices[i].device > maxDevice ) {
1079             maxDevice = devices[i].device;
1080         }
1081     }
1082 
1083     // Create the global context.
1084     cnmem::Context::create();
             ...
```

cnmemInit() calls cnmem::Context::create() in line 1084 and

```
1024 cnmemStatus_t Context::create() {
1025     sCtx = new Context;
1026     sCtxCheck = CTX_VALID;
1027     return CNMEM_STATUS_SUCCESS;
1028 }
```

create() resets the Context class's static member variable sCtx to a newly created Context object in sCtx.

So, if I call rmmInitialize() multiple times (after cudaSetDevice(), once per device), only the last call will have effect (besides memory leaks for previously allocated Context objects).

rmmInitialize() does not take num_devices as cnmemInit, so I cannot initialize RMM for multiple devices in a single rmmInitialize() call, either.

**Describe the solution you'd like**
Need a mechanism to initialize RMM for multiple devices (in cnmem style or by calling rmmInitialize multiple times after cudaSetDevice).

**Comment by jrhemstad** [comment]
I propose that `rmmInitialize` just be updated to accept an optional `std::vector<int> devices` that will be forwarded to `cnmemInit`. 

It will still be the burden of the user to call `cudaSetDevice` before allocating/deallocating any memory.

**Comment by seunghwak** [comment]
> I propose that `rmmInitialize` just be updated to accept an optional `std::vector<int> devices` that will be forwarded to `cnmemInit`.

Yes, we may update rmmOptions_t to take std::vector<int> devices.

> It will still be the burden of the user to call cudaSetDevice before allocating/deallocating any memory. Or, we could add a device argument to device_memory_resource::allocate.

Either is fine for me as long as there is a mechanism to control target GPU device to allocate memory.

**Comment by jrhemstad** [comment]
Looking at what Thrust did for their `memory_resource`, they maintain a separate resource for each device: https://github.com/thrust/thrust/blob/63d847beab9931978d9c894afd7432a6f94abd46/thrust/system/cuda/detail/per_device_resource.h

With `rmm::mr::memory_resource`, I think it makes sense to emulate this.

**Comment by jrhemstad** [comment]
I've added initial support for multi-GPU RMM here: https://github.com/rapidsai/rmm/pull/106

Several caveats:
- This is a short-term fix to unblock mutli-GPU work with RMM
- It introduces changes that make it incompatible with RMM's current CFFI Python bindings, and as such, the Python bindings will not work with this PR
- It is intended for use at the C/C++ layer
- I reserve the right to change the API significantly

**Comment by harrism** [comment]
Fixed by #106
